### ISSUE TITLE
Multi-Grad Hooks

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#83617]
### 🚀 The feature, motivation and pitch

For FSDP, one advantage of the `FlatParameter` is that registering a hook on its `AccumulateGrad` object gives us the correct time to schedule a reduce-scatter: The hook runs when all constituent original parameters' gradients are ready.

An alternative to this paradigm is to have a "multi-grad hook" that only runs once all passed-in tensors' gradients are ready. In that case, we can explore a new design for FSDP where we are not constrained to a single (large) `flat_param.grad`. Under such a new design, the `FlatParameter` would **only be for communication** and no longer relevant for autograd.

One effort this half is to provide an option to expose the original parameters in FSDP. The current approach manually sets `.data` to be views into the `FlatParameter`, and another approach is to use a tensor subclass like `IndirectParameter`. Regardless of the approach, as long as we have the `FlatParameter` interface with autograd and thus have its `.grad` attribute populated, we cannot save memory when only _some_ of the original parameters require gradients. The parts of the `.grad` corresponding to parameters that do not require gradients will simply be zeros, but the entire `.grad` needs to be materialized.

This use case surfaces in finetuning, where there is a large model with only some small subset of parameters that require gradients and those parameters are spread throughout the module tree. PyTorch FSDP does not have good support for this at the moment. One option is to support a module having multiple `FlatParameter`s, where we have the ones that do require gradient be in the same smaller `FlatParameter` and the ones that do not require gradient in another larger `FlatParameter`. This can work but is not ideal since it (unexpectedly) couples communication and autograd. The smaller `FlatParameter` will be less efficient to communicate.

### Alternatives

The alternative is the current approach of having the `FlatParameter` interface with autograd.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by albanD** [comment]
Does the following work for you?
```python
import torch

def register_multi_grad_hook(tensors, fn):
    count = 0
    nb_calls = len(tensors)
    buffer = [None] * nb_calls

    def get_inner_hook(idx):
        def inner_hook(grad):
            nonlocal count
            buffer[idx] = grad
            count += 1

            if count == nb_calls:
                fn(buffer)
        return inner_hook
    for i, t in enumerate(tensors):
        t.register_hook(get_inner_hook(i))


t1 = torch.rand(2, requires_grad=True)
t2 = torch.rand(2, requires_grad=True)
t3 = torch.rand(2, requires_grad=True)
t4 = torch.rand(2, requires_grad=True)

def hook(grads):
    print(f"Multi-hook called with {len(grads)} gradients")
register_multi_grad_hook((t2, t3), hook)

def get_hook(name):
    def hook(grad):
        print(f"{name} hook called")
    return hook
t1.register_hook(get_hook("t1"))
t3.register_hook(get_hook("t3"))


out = t1.clone()
out = out + t2
out = out + t3
out = out + t4

out.sum().backward()



```
This will output
```
t3 hook called
Multi-hook called with 2 gradients
t1 hook called

```

**Comment by ezyang** [comment]
No because the multihook is not called if some of the hooks are unused

**Comment by albanD** [comment]
From offline discussion:
- There is no way to do without this at python level
- There are two main avenues to do this:
  - We can change the engine to do it:
    - Add a new special kind of hook that the engine knows about
    - We can discover if/when it needs to be called "for free" by re-using the traversal that we already do to compute dependencies_
    - This new state will need to be stored on graph_task to ensure that multiple backward can properly run in parallel
  - Create a custom Node that is always needed_ if linked to the graph.
    - This hook will just be another Node in the graph
    - Will need to allow for an edge to have multiple endpoint <- major change

**Comment by soulitzer** [comment]
I created https://github.com/pytorch/pytorch/pull/84773 for a different design, which aims to make fewer modifications to the engine/graph itself, but may require more changes on the Python side. It does the job by exposing a way to query the TLS graph task's exec_info which is a map mapping the Node to a bool indicating whether it will be executed in the current backward pass (as determined by the inputs= argument for .grad of .backward).

Below, I've modified Alban's example above using the new API to address Ed's comment.
```python
import torch

def register_multi_grad_hook(tensors, fn):
    count = 0
    nb_calls = None
    buffer = None

    def get_grad_fn(t):
        # or grad accumulator
        if t.requires_grad and t.grad_fn is None:
            return t.clone().grad_fn.next_functions[0][0]
        else:
            return t.grad_fn

    grad_fns = list(map(get_grad_fn, tensors))

    def get_inner_hook(idx):
        def inner_hook(grad):
            nonlocal count, nb_calls, buffer

            if count == 0:
                # On the first call, compute the actual nb_calls and buffer
                nb_calls = sum(1 for g in grad_fns if torch._C._will_engine_execute_node(g))
                buffer = [None] * nb_calls

            buffer[idx] = grad
            count += 1

            if count == nb_calls:
                fn(buffer)
        return inner_hook
    for i, t in enumerate(tensors):
        t.register_hook(get_inner_hook(i))


t1 = torch.rand(2, requires_grad=True)
t2 = torch.rand(2, requires_grad=True)
t3 = torch.rand(2, requires_grad=True)
t4 = torch.rand(2, requires_grad=True)

def hook(grads):
    print(f"Multi-hook called with {len(grads)} gradients")
register_multi_grad_hook((t2, t3), hook)

def get_hook(name):
    def hook(grad):
        print(f"{name} hook called")
    return hook
t1.register_hook(get_hook("t1"))
t3.register_hook(get_hook("t3"))


out = t1.clone()
out = out + t2
out = out + t3
out = out + t4

out.sum().backward(inputs=(t2, t3))
# t3 hook called
# Multi-hook called with 2 gradients
# t1 hook called

out.sum().backward(inputs=(t1, t2))
# Multi-hook called with 1 
```

**Comment by ezyang** [comment]
This alternate API seems fine too!

**Comment by ezyang** [comment]
I think the primary risk is if this adds overhead, but @awgu can probably test this

**Comment by awgu** [comment]
Thanks @soulitzer for prototyping this so quickly! I will find some time to try it out soon.

**Comment by soulitzer** [comment]
Hey @awgu, we're landing https://github.com/pytorch/pytorch/pull/86260, which is a cleaned up version of https://github.com/pytorch/pytorch/issues/83617#issuecomment-1244048429. One improvement is that it now support multiple backwards on different threads on the same graph.

One thing to note is that if an error happens during backward the state of the hook is still kept alive by the graph, so that can be problematic if someone catches the error and runs backward on the same graph. We may find a more automatic solution later, but for now to get around this one should just remove and reregister the hook to reset its state (this would reset the state for all backwards running concurrently).

Let us know how it works for you!

**Comment by awgu** [comment]
Hi @soulitzer. Thanks so much for pushing this forward! Sadly, from the FSDP side, there have been a series of blockers that have emerged between our current state and adoption of multi-grad hooks. Integrating multi-grad hooks will not be in scope for this half :/
